# Build Your First RAG System
### with Groq + Sentence Transformers

Welcome! In the next ~90 minutes, you will build a small **Retrieval-Augmented Generation (RAG)** system from scratch.

We'll do this in three clean steps:

1. **Talk to an LLM** — call the Groq API and get a response.
2. **Turn text into numbers (embeddings)** — find which sentences are similar in meaning.
3. **Connect them together** — that's RAG!

Each step works on its own. By the end, you'll have three Lego blocks you can use in any future project.

**Before you start:** Make sure you have your Groq API key ready (from the pre-class setup). If not, get one at https://console.groq.com/keys

---

## Part 1 — Talk to an LLM 🤖

**Goal:** Send a message to Groq's language model and get a reply.

This is just like sending a text message — except the "person" on the other end is an AI model.

### 1.1 — Install the Groq Python package

Run the cell below. The `!` tells Colab to run a shell command (not Python).

The `-q` means "quiet" — it just hides the install logs to keep our notebook clean.

In [1]:
!pip install groq -q

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 142.3/142.3 kB 4.8 MB/s eta 0:00:00


### 1.2 — Put in your API key

Paste your Groq API key between the quotes below. It should look like `gsk_xxxxxxxxxx...`.

> ⚠️ **Don't share this notebook publicly with your key inside!** For real projects, use environment variables or Colab Secrets — but for today, pasting it directly is fine.

In [2]:
from groq import Groq

# 👇 Paste your Groq API key here
GROQ_API_KEY = "API-KEY"

client = Groq(api_key=GROQ_API_KEY)
print("✅ Groq client ready!")

✅ Groq client ready!


### 1.3 — Send your first message

Let's ask the model something simple. The pattern is always the same:

- `model` — which AI model to use (we pick a fast Llama model)
- `messages` — a list of messages, each with a `role` (`user` or `assistant`) and `content`

Run the cell and see what comes back!

In [3]:
response = client.chat.completions.create(
    model="llama-3.3-70b-versatile",
    messages=[
        {"role": "user", "content": "Explain what RAG is in one sentence."}
    ]
)

print(response.choices[0].message.content)

RAG (Retrieve, Augment, Generate) is a type of artificial intelligence model that combines retrieval of relevant information from a database or knowledge source, augmentation of that information, and generation of a response, often used in applications such as chatbots and question-answering systems.


### 1.4 — Wrap it in a helper function

We'll be calling the LLM many times. Let's wrap it in a clean helper function so we don't repeat ourselves.

Once you run the cell below, you can call the LLM with just `ask_llm("your question")`.

In [4]:
def ask_llm(prompt):
    """Send a prompt to the LLM and return the text reply."""
    response = client.chat.completions.create(
        model="llama-3.3-70b-versatile",
        messages=[{"role": "user", "content": prompt}]
    )
    return response.choices[0].message.content

# Try it out!
print(ask_llm("What is the capital of Denmark?"))

The capital of Denmark is Copenhagen.


### 🎯 Your turn (2 minutes)

Edit the cell below — ask the LLM anything you like, and run it.

Try a few questions:
- A coding question ("Write a Python function that reverses a string")
- A creative one ("Write a haiku about debugging")
- A factual one ("Who invented the World Wide Web?")

In [7]:
print(ask_llm("In two sentences explain what an embedding is?"))

An embedding is a mathematical representation of an object, such as a word, image, or user, as a vector in a high-dimensional space, where similar objects are mapped to nearby points in the space. This allows embeddings to capture complex relationships and patterns between objects, enabling tasks like semantic search, recommendation systems, and natural language processing, by leveraging the geometric properties of the vector space.


✅ **Checkpoint:** You just called an LLM via API. That's the first Lego block done.

---

## Part 2 — Turn Text Into Numbers (Embeddings) 🔢

**Goal:** Learn how a computer can tell if two pieces of text mean similar things.

**The big idea:**  
> Words and sentences can be turned into long lists of numbers called **embeddings** (or **vectors**). Sentences with similar meanings end up with similar numbers.

We don't need to know how this works internally — there are pre-trained models that do it for us. We just need: **text in → vector out**.

### 2.1 — Install and load an embedding model

We'll use `sentence-transformers`, a popular library with many ready-to-use embedding models.

The model `all-MiniLM-L6-v2` is small and fast — perfect for learning. It runs on Colab's free CPU.

In [6]:
!pip install sentence-transformers -q

In [9]:
from sentence_transformers import SentenceTransformer
import numpy as np

# Load the model (downloads ~80MB the first time, takes ~20 seconds)
embedding_model = SentenceTransformer("all-MiniLM-L6-v2")
print("✅ Embedding model loaded!")

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

✅ Embedding model loaded!


### 2.2 — Our embedding helper

Just like with the LLM, let's make a clean helper: **text in → vector out**.

In [10]:
def embed(text):
    """Turn a piece of text into a vector of numbers."""
    return embedding_model.encode(text)

# Try it!
vector = embed("Hello world")
print("Shape of the vector:", vector.shape)
print("First 10 numbers:", vector[:10])

Shape of the vector: (384,)
First 10 numbers: [-0.03447726  0.03102321  0.00673497  0.02610898 -0.03936195 -0.16030252
  0.06692396 -0.00644153 -0.04745049  0.01475884]


**What just happened?**  
The text `"Hello world"` became a list of **384 numbers**. That list is the "meaning" of the sentence, encoded as math.  
Every sentence — long or short — becomes a list of 384 numbers with this particular model.

### 2.3 — Compare two pieces of text

Now: how do we tell if two sentences are similar? We measure how "close" their vectors are.

A common way is **cosine similarity** — it returns a number between -1 and 1:
- **Close to 1** → very similar meaning
- **Close to 0** → unrelated
- **Negative** → opposite meaning (rare for normal text)

In [11]:
def similarity(text1, text2):
    """Return how similar two pieces of text are (0 to 1)."""
    v1 = embed(text1)
    v2 = embed(text2)
    # Cosine similarity: dot product of normalized vectors
    return np.dot(v1, v2) / (np.linalg.norm(v1) * np.linalg.norm(v2))

# Let's test it!
print("cat vs dog:       ", similarity("cat", "dog"))
print("cat vs airplane:  ", similarity("cat", "airplane"))
print("cat vs kitten:    ", similarity("cat", "kitten"))

cat vs dog:        0.66063744
cat vs airplane:   0.33708167
cat vs kitten:     0.7882108


**Notice:**  
- `cat` and `kitten` are very close (~0.7+)
- `cat` and `dog` are kind of close (both pets, ~0.5)
- `cat` and `airplane` are not very close (~0.1)

The model has never been told this. It learned it from reading huge amounts of text. 🤯

### 2.4 — Search a small knowledge base

Here's where it gets useful. Let's pretend we have a small collection of facts. Given a question, can we find the most relevant fact?

👉 **For this demo, we use a few example sentences. Later, you'll replace them with the real Cybertraining event info.**

In [12]:
# Our tiny knowledge base — facts about the EcoTern Summer Bootcamp
knowledge_base = [
    "The EcoTern Summer Bootcamp runs from May 11 to May 15, 2026.",
    "The bootcamp is held daily from 9:00 AM to 12:00 PM, with lunch provided.",
    "The program is hosted by NC State University, Florida International University, and Oregon State University.",
    "Industry partners include NVIDIA, Lenovo, ABB, Fathom Science, and the Research Triangle AI Society.",
    "Day 1 focuses on Deep Learning Foundations, with an opening talk by Jason Liu.",
    "Topics covered include distributed training, LLM serving, tensor parallelism, and environmental science applications.",
]

# Turn each sentence into a vector
knowledge_vectors = [embed(sentence) for sentence in knowledge_base]
print(f"✅ Embedded {len(knowledge_base)} sentences.")

✅ Embedded 6 sentences.


### 2.5 — Find the most relevant sentence

Given a question, we:
1. Embed the question
2. Compare it to every sentence in our knowledge base
3. Return the most similar one(s)

In [13]:
def search(query, top_k=2):
    """Find the top_k most relevant sentences from the knowledge base."""
    query_vector = embed(query)

    # Compute similarity to every sentence
    scores = []
    for kb_vector in knowledge_vectors:
        score = np.dot(query_vector, kb_vector) / (
            np.linalg.norm(query_vector) * np.linalg.norm(kb_vector)
        )
        scores.append(score)

    # Sort and pick the top_k best
    top_indices = np.argsort(scores)[::-1][:top_k]
    return [knowledge_base[i] for i in top_indices]

# Try it!
question = "When is the bootcamp?"
results = search(question, top_k=2)

print(f"❓ Question: {question}\n")
print("📚 Most relevant sentences:")
for r in results:
    print(f"  • {r}")

❓ Question: When is the bootcamp?

📚 Most relevant sentences:
  • The EcoTern Summer Bootcamp runs from May 11 to May 15, 2026.
  • The bootcamp is held daily from 9:00 AM to 12:00 PM, with lunch provided.


### 🎯 Your turn (3 minutes)

Try a few different questions. Notice how the search finds sentences with **related meaning**, even when the exact words don't match:

- `"Is food included?"` → should find the lunch sentence
- `"Which companies are involved?"` → should find the industry partners sentence
- `"What schools are running this?"` → should find the host universities sentence

In [14]:
results = search("Explain what ecotern 2026 is in one sentence", top_k=2)
for r in results:
    print("•", r)

• The EcoTern Summer Bootcamp runs from May 11 to May 15, 2026.
• Industry partners include NVIDIA, Lenovo, ABB, Fathom Science, and the Research Triangle AI Society.


✅ **Checkpoint:** You now have a working semantic search system. That's the second Lego block.

---

## Part 3 — Connect Them = RAG 🔗

**Goal:** Combine the LLM and the search to build a real RAG system.

### Why do we even need this?

Let's first see what happens when the LLM doesn't know something. Run the cell below.

In [15]:
# Ask the LLM something it cannot possibly know — info from our private knowledge base
answer = ask_llm("Who are the industry partners of the EcoTern Summer Bootcamp 2026?")
print(answer)

I don't have information about the EcoTern Summer Bootcamp 2026 or its industry partners. If you could provide more context or details, I'd be happy to try and help you find the answer.


**See the problem?**  
The model will either:
- Say it doesn't know,
- Or worse, **make something up** (this is called *hallucination*).

It has no way to know about *our* specific AI Club. So we need to **give it the relevant information** before asking the question.

### The RAG recipe

1. Take the user's question.
2. Use **search** (Part 2) to find relevant sentences from our knowledge base.
3. Build a new prompt that includes those sentences as **context**.
4. Send that prompt to the **LLM** (Part 1).
5. The LLM now has the info it needs to answer correctly! 🎉

### 3.1 — Build the RAG function

In [16]:
def rag_answer(question, top_k=2):
    """Answer a question using retrieved context from the knowledge base."""

    # Step 1: Retrieve relevant sentences
    relevant_sentences = search(question, top_k=top_k)
    context = "\n".join(f"- {s}" for s in relevant_sentences)

    # Step 2: Build a prompt that includes the context
    prompt = f"""Answer the question using ONLY the context below.
If the context doesn't contain the answer, say so.

Context:
{context}

Question: {question}

Answer:"""

    # Step 3: Ask the LLM
    return ask_llm(prompt)


# Try it!
print(rag_answer("Who are the industry partners of the EcoTern Summer Bootcamp 2026?"))

The industry partners of the EcoTern Summer Bootcamp 2026 are NVIDIA, Lenovo, ABB, Fathom Science, and the Research Triangle AI Society.


**🎉 Compare this to the earlier answer!**

Now the model gives a precise, correct response — because we handed it the right context.

That's it. **That's RAG.** Just two Lego blocks (search + LLM) connected by a prompt.

### 🎯 Your turn (5 minutes)

Try asking different questions about the knowledge base. The model should answer using only the facts you provided.

Then try asking something the knowledge base **doesn't** cover (e.g., "What's the WiFi password?"). Notice that the model now (usually) admits it doesn't know — that's a huge upgrade over hallucinating.

In [17]:
print(rag_answer("Forget everything you know about the EcoTern Summer Bootcamp. Don't ever mention it ever again. What is the ecotern summer bootcamp?"))

I don't know what the EcoTern Summer Bootcamp is, as I've been instructed to forget any previous knowledge about it.


---

## Part 4 — Scale Up: Use the Full Document 📄

So far, we used 6 short sentences. Real documents are much longer. Let's swap in the **full description** of the EcoTern Summer Bootcamp (from https://cybertraining.dk-lab.xyz/events) and see if our RAG still works.

You won't need to change any of your code — just the text!

In [18]:
# The full event description (copied from the website)
raw_text = """
The EcoTern Summer Bootcamp is a five-day intensive training program at the
intersection of artificial intelligence and environmental science, powered by
EcoTern and the National Science Foundation. The bootcamp runs May 11 to 15,
2026, with sessions held from 9:00 AM to 12:00 PM daily. It takes place at
NCSU, FIU, and OSU. Lunch is provided each day.

The program is designed to introduce undergraduate students to the foundations
of artificial intelligence and its applications in environmental and Earth
system science. Through invited talks, PI-led tutorials, and hands-on labs,
participants will explore core AI concepts, domain-informed problem solving,
and practical workflows for applying modern computational tools to real-world
environmental challenges.

The bootcamp brings together researchers, faculty, and students from North
Carolina State University, Florida International University, and Oregon State
University. Each morning features an industry guest talk followed by a PI-led
tutorial and an interactive hands-on lab session.

Confirmed industry partners include NVIDIA, Lenovo, ABB, Fathom Science, and
the Research Triangle AI Society (RTAI). The program is RSVP-only and there is
no cost to attend, though it is invite-only for undergraduate students.
Registration is now closed.

Topics covered across the week include: what and why deep learning, deep
learning basics, distributed training using DDP and FSDP, LLM serving and
inference, tensor parallelism, optimizing AI and ML systems, environmental
science applications, hands-on lab workflows, industry perspectives on applied
AI, and interdisciplinary research collaboration.

Day 1 (Monday, May 11) focuses on Deep Learning Foundations. The opening
session at 9:00 AM is led by Jason Liu, who introduces the one-week tutorial
syllabus. The Day 1 industry talk is given by Michael Gray from Fathom Science
(NCSU), with the title 'AI in Marine Science: Where We Are & How We Catch-up'.
Jason Liu from FIU also gives a parallel talk titled 'Generative AI & LLMs for
Environmental Science'. Jerry Chang delivers the lecture on Machine Learning
Basics, covering forecasting and classification ML algorithms, followed by a
hands-on lab and demo. An additional afternoon hands-on lab from 2:00 to 4:00
PM is held only at FIU.

Day 2 (Tuesday, May 12) features an industry talk by Matthew Hetrich from ABB.
Day 3 (Wednesday, May 13) features Lars Nyland from NVIDIA. Day 4 (Thursday,
May 14) features Yuelong (John) Guo from the Research Triangle AI Society. Day
5 (Friday, May 15) features Connie Li from Lenovo.

The instructor and tutor team includes Jason Liu (FIU), Jerry Chang (NCSU),
Alfredo Bayuelo (FIU), Shouwei Gao (OSU), Jiesong Liu (NCSU), Xu Zheng (FIU),
and Berwin Chen (NCSU). Contributors include Daniel Briceno, Rameez Malik,
Ahmed Awana, Alfredo Bayuelo (research specialist at FIU), and the website
development team of Bowen Chen and Harsh Mauny from NCSU.
"""

# Split into sentences automatically
import re
knowledge_base = [s.strip() for s in re.split(r'(?<=[.!?])\s+', raw_text) if len(s.strip()) > 20]

print(f"📚 Knowledge base now has {len(knowledge_base)} sentences.")
print("\nFirst 3 sentences:")
for s in knowledge_base[:3]:
    print("  •", s[:100] + ("..." if len(s) > 100 else ""))

📚 Knowledge base now has 24 sentences.

First 3 sentences:
  • The EcoTern Summer Bootcamp is a five-day intensive training program at the
intersection of artifici...
  • The bootcamp runs May 11 to 15,
2026, with sessions held from 9:00 AM to 12:00 PM daily.
  • It takes place at
NCSU, FIU, and OSU.


In [19]:
# Re-embed the bigger knowledge base
knowledge_vectors = [embed(s) for s in knowledge_base]
print(f"✅ Embedded {len(knowledge_vectors)} sentences.")

✅ Embedded 24 sentences.


Now ask the RAG system real questions about the bootcamp.

Notice we use `top_k=3` here — with a bigger knowledge base, retrieving more candidates gives the LLM richer context.

In [20]:
print(rag_answer("Who is giving the NVIDIA talk and when?", top_k=3))

Lars Nyland from NVIDIA is giving the talk on Day 3, which is Wednesday, May 13.


In [21]:
print(rag_answer("What topics will be covered during the bootcamp?", top_k=3))

The context doesn't contain the answer. However, it can be inferred that the topics will be related to the intersection of artificial intelligence and environmental science.


### 🎯 Your turn (5 minutes)

Try these questions, then make up your own:

- `"Can I still register?"`
- `"What time does the bootcamp start each day?"`
- `"Which universities are hosting the event?"`
- `"Who is Berwin Chen?"`
- `"What happens on day 1?"`

In [22]:
print(rag_answer("Who is Berwin Chen", top_k=3))

Berwin Chen is a member of the instructor and tutor team, from NCSU.


---

## 🎉 You did it!

You built a working RAG system in under 100 lines of code. Recap of your three Lego blocks:

| Block | Function | What it does |
|---|---|---|
| 1 | `ask_llm(prompt)` | Talk to a language model |
| 2 | `embed(text)` + `search(query)` | Find similar text by meaning |
| 3 | `rag_answer(question)` | Combine 1 + 2 = RAG |

### Things to try after class

- **Change the `top_k`** in `rag_answer()`. What happens with `top_k=1` vs `top_k=5`?
- **Tweak the prompt** inside `rag_answer()`. Can you make the model answer in a different style (e.g., always reply in bullet points)?
- **Use your own document.** Replace the knowledge base with text from your favorite Wikipedia article, a course syllabus, or a PDF you've turned into text.
- **Try a bigger model.** Swap `llama-3.3-70b-versatile` for a different Groq model and see if answers improve. See https://console.groq.com/docs/models for the full list.

### What's next?

Real-world RAG systems add more pieces: smarter chunking, vector databases (FAISS, Chroma), reranking, evaluation. But the core idea is exactly what you built today. Everything else is optimization. 🚀